ProjectSource/app-library/src/test/java/com/kayako/sdk/android/k5/messenger/messagelistpage/helpers/OffboardingHelperTest.java

```diff
@@ -0,0 +1,335 @@
+        // Arrange
+        String feedback = "good";
+        // Act
+        invokeMethod(underTest, "addOrUpdateRating", offboardingHelperViewCallback, Rating.SCORE
+                .GOOD, feedback);
+        // Assert
+        ArgumentCaptor<Rating.SCORE> argumentCaptor1 = ArgumentCaptor.forClass(Rating.SCORE.class);
+        ArgumentCaptor<String> argumentCaptor2 = ArgumentCaptor.forClass(String.class);
+        verify(offboardingHelperViewCallback).onAddRating(argumentCaptor1.capture(),
+                argumentCaptor2.capture());
+        assertEquals(Rating.SCORE.GOOD, argumentCaptor1.getValue());
+        assertEquals(feedback, argumentCaptor2.getValue());
+    }
+
+    @Test
+    public void givenAddOrUpdateRatingWhenFeedbackNullThenOnUpdateRating() throws Exception {
+        // Arrange
+        String feedback = null;
+        AtomicLong mIdOfLastRatingAdded = new AtomicLong();
+        mIdOfLastRatingAdded.set(42L);
+        setInternalState(underTest, "mIdOfLastRatingAdded", mIdOfLastRatingAdded);
+        // Act
+        invokeMethod(underTest, "addOrUpdateRating", offboardingHelperViewCallback, Rating.SCORE
+                .GOOD, feedback);
+        // Assert
+        ArgumentCaptor<Long> argumentCaptor1 = ArgumentCaptor.forClass(Long.class);
+        ArgumentCaptor<Rating.SCORE> argumentCaptor2 = ArgumentCaptor.forClass(Rating.SCORE.class);
+        verify(offboardingHelperViewCallback).onUpdateRating(argumentCaptor1.capture(),
+                argumentCaptor2.capture());
+        assertEquals(Long.valueOf(42L), argumentCaptor1.getValue());
+        assertEquals(Rating.SCORE.GOOD, argumentCaptor2.getValue());
+    }
+
+    @Test
+    public void givenAddOrUpdateRatingWhenFeedbackNotNullAndLastAddedIdNotZeroThenOnUpdateRating
+            () throws Exception {
+        // Arrange
+        String feedback = "feedback";
+        AtomicLong mIdOfLastRatingAdded = new AtomicLong();
+        mIdOfLastRatingAdded.set(42L);
+        setInternalState(underTest, "mIdOfLastRatingAdded", mIdOfLastRatingAdded);
+        // Act
+        invokeMethod(underTest, "addOrUpdateRating", offboardingHelperViewCallback, Rating.SCORE
+                .GOOD, feedback);
+        // Assert
+        ArgumentCaptor<Long> argumentCaptor1 = ArgumentCaptor.forClass(Long.class);
+        ArgumentCaptor<Rating.SCORE> argumentCaptor2 = ArgumentCaptor.forClass(Rating.SCORE.class);
+        ArgumentCaptor<String> argumentCaptor3 = ArgumentCaptor.forClass(String.class);
+        verify(offboardingHelperViewCallback).onUpdateFeedback(argumentCaptor1.capture(),
+                argumentCaptor2.capture(), argumentCaptor3.capture());
+        assertEquals(Long.valueOf(42L), argumentCaptor1.getValue());
+        assertEquals(Rating.SCORE.GOOD, argumentCaptor2.getValue());
+        assertEquals("feedback", argumentCaptor3.getValue());
+    }
+
+    @Test
+    public void givenAddToQueue() throws Exception {
+        // Arrange
+        String feedback = "good";
```

Comment:
⚠️  This snippet of code contains local variables that have a "constant" behavior - they are **effectively final**. Consider marking these local variables as `final` to enforce their "constant" behavior and take advantage of the compiler support if we try to intentionally/accidentally modify these variables.  

**[More details](https://confluence.devfactory.com/x/uwkGF)**
          
## The possible solution in this case

```JAVA
final String feedback = "good";

```
 

<a id="c619d8a0-5328-11e8-a476-4b3e34747147" href="none" codereview-uuid-reserved-tag />